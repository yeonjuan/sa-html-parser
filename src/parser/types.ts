export enum InsertionMode {
  initialMode,
  afterInitialMode,
  // beforeHtmlMode,
  // beforeHeadMode,
  // inHeadMode,
  // inHeadNoScriptMode,
  // afterHeadMode,
  // inBodyMode,
  // textMode,
  // inTableMode,
  // inTableTextMode,
  // inCaptionMode,
  // inColumnGroupMode,
  // inTableBodyMode,
  // inRowMode,
  // inCellMode,
  // inSelectMode,
  // inSelectInTableMode,
  // inTemplateMode,
  // afterBodyMode,
  // inFramesetMode,
  // afterFramesetMode,
  // afterAfterBodyMode,
  // afterAfterFramesetMode,
}
