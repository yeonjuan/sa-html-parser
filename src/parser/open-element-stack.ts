export class OpenElementStack {
  public readonly elements: any[] = [];
  public stackTop = -1;
  public top: any = null;

  public push(element: any) {
    this.top = element;
    this.stackTop++;
    this.elements[this.stackTop] = element;
  }

  public pop() {
    this.stackTop--;
    this.top = this.elements[this.stackTop];
  }

  public popUntilElementPopped(element: any): any[] {
    let elements: any[] = [];
    while (this.stackTop > -1) {
      const poppedElement = this.top;
      elements.push(poppedElement);
      this.pop();
      if (poppedElement === element) {
        break;
      }
    }
    return elements;
  }
}
