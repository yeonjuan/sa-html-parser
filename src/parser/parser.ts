import {
  CommentNode,
  DoctypeNode,
  RootNode,
  AnyNode,
  ElementNode,
  ClosingElementNode,
  TextNode,
} from "../nodes";
import { Tokenizer } from "../tokenizer/tokenizer";
import {
  AnyHtmlToken,
  AttributeToken,
  CharacterLikeToken,
  CommentToken,
  DoctypeToken,
  EndTagToken,
  EofToken,
  HtmlTokenType,
  NullToken,
  StartTagToken,
} from "../tokens";
import { OpenElementStack } from "./open-element-stack";
import * as utils from "../common/utils";

export class Parser {
  private root!: RootNode;
  private html!: string;
  private tokenizer!: Tokenizer;
  private openElementStack = new OpenElementStack();
  private running = true;

  constructor() {}

  public parse(html: string): RootNode {
    this.html = html;
    this.tokenizer = Tokenizer.create(this.html);
    this.createRoot();

    while (this.running) {
      const token = this.tokenizer.getNextToken();
      if (!token) {
        break;
      }
      this.process(token);
    }

    return this.root;
  }

  private createRoot() {
    this.root = new RootNode();
    this.openElementStack.push(this.root);
  }

  private insertCommentNodeToRoot(node: CommentNode) {
    this.insertToCurrent(node);
    this.root.comments.push(node);
  }

  private insertToCurrent(node: AnyNode) {
    this.openElementStack.top.children.push(node);
  }

  private pushToOpenStack(node: any) {
    this.openElementStack.push(node);
  }

  private popFromOpenStackUntilTagName(tagName: string): null | any[] {
    let unclosedElements: null | any[] = null;
    for (let i = this.openElementStack.stackTop; i >= 0; i--) {
      const element = this.openElementStack.elements[i];
      if (element.tagName === tagName) {
        unclosedElements = this.openElementStack.popUntilElementPopped(element);
        break;
      }
    }
    return unclosedElements;
  }

  private process(token: AnyHtmlToken) {
    this[token.type]?.(token as any);
  }

  private [HtmlTokenType.Comment](token: CommentToken) {
    const comment = CommentNode.fromToken(token);
    this.insertToCurrent(comment);
    this.insertCommentNodeToRoot(comment);
  }

  private [HtmlTokenType.Doctype](token: DoctypeToken) {
    this.insertToCurrent(DoctypeNode.fromToken(token));
  }

  private [HtmlTokenType.StartTag](token: StartTagToken) {
    const tagNode = ElementNode.fromToken(token);
    this.insertToCurrent(tagNode);
    if (token.selfClosing) {
      tagNode.selfClosing = true;
    } else {
      this.pushToOpenStack(tagNode);
    }
  }

  private [HtmlTokenType.EndTag](token: EndTagToken) {
    const closing = ClosingElementNode.fromToken(token);
    const poppedElements = this.popFromOpenStackUntilTagName(
      token.tagName.value
    );
    if (!poppedElements) {
      throw new Error();
    }
    const element = utils.last<ElementNode>(poppedElements)!;
    element.children = utils.getChildrenRecursively(element);
    element.closingElement = closing;
    element.end = closing.end;
    element.loc.end = closing.loc.end;
    element.range[1] = closing.range[1];
  }

  private [HtmlTokenType.CharacterLike](token: CharacterLikeToken) {
    if (this.openElementStack.top.children?.length) {
      const lastChild: any = utils.last(this.openElementStack.top.children);
      if (lastChild.type === "Text") {
        const textNode = lastChild as TextNode;
        textNode.value += token.value.value;
        textNode.end = token.end;
        textNode.loc.end = token.value.loc.end;
        return;
      }
    }
    this.insertToCurrent(TextNode.fromToken(token));
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private [HtmlTokenType.EOF](token: EofToken) {
    const poppedElements = this.openElementStack.popUntilBeforeElementPopped(
      this.root
    );
    if (poppedElements.length) {
      for (let i = poppedElements.length - 2; i >= 0; i--) {
        this.root.children.push(poppedElements[i]);
      }
    }
    this.running = false;
    const lastChild = utils.last<AnyNode>(this.root.children);
    this.root.range[1] = lastChild?.range[1] || 0;
    this.root.end = lastChild?.range[1] || 0;
    this.root.loc.end = lastChild?.loc.end || {
      column: 0,
      line: 1,
    };
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private [HtmlTokenType.Null](token: NullToken) {}
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private [HtmlTokenType.Attribute](token: AttributeToken) {}
}
