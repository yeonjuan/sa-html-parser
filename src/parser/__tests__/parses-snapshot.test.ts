import { parse } from "../parser";

describe("parses: snapshot", () => {
  test("basic", () => {
    const result = parse(
      `<span class="_1syGnXOL _3di88A4c" data-clk="dropbanner1b" style="padding-right: 20px; color: white; padding-left: 20px"><span>웨일 사용자 목소리입니다.</span></span>`
    );
    expect(result).toMatchSnapshot();
  });
});
