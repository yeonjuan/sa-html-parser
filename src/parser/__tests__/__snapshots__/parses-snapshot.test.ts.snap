// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses: snapshot basic 1`] = `
Root {
  "children": Array [
    ElementNode {
      "attrs": Array [
        AttributeNode {
          "end": 33,
          "loc": Object {
            "end": Object {
              "column": 33,
              "line": 1,
            },
            "start": Object {
              "column": 6,
              "line": 1,
            },
          },
          "name": AttrNameToken {
            "end": 11,
            "loc": Object {
              "end": Object {
                "column": 11,
                "line": 1,
              },
              "start": Object {
                "column": 6,
                "line": 1,
              },
            },
            "range": Array [
              6,
              11,
            ],
            "start": 6,
            "type": "AttributeName",
            "value": "class",
          },
          "range": Array [
            6,
            33,
          ],
          "start": 6,
          "type": "Attribute",
          "value": AttrValueToken {
            "end": 33,
            "loc": Object {
              "end": Object {
                "column": 33,
                "line": 1,
              },
              "start": Object {
                "column": 12,
                "line": 1,
              },
            },
            "range": Array [
              12,
              33,
            ],
            "start": 12,
            "type": "AttributeValue",
            "value": "\\"_1syGnXOL _3di88A4c\\"",
          },
        },
        AttributeNode {
          "end": 57,
          "loc": Object {
            "end": Object {
              "column": 57,
              "line": 1,
            },
            "start": Object {
              "column": 34,
              "line": 1,
            },
          },
          "name": AttrNameToken {
            "end": 42,
            "loc": Object {
              "end": Object {
                "column": 42,
                "line": 1,
              },
              "start": Object {
                "column": 34,
                "line": 1,
              },
            },
            "range": Array [
              34,
              42,
            ],
            "start": 34,
            "type": "AttributeName",
            "value": "data-clk",
          },
          "range": Array [
            34,
            57,
          ],
          "start": 34,
          "type": "Attribute",
          "value": AttrValueToken {
            "end": 57,
            "loc": Object {
              "end": Object {
                "column": 57,
                "line": 1,
              },
              "start": Object {
                "column": 43,
                "line": 1,
              },
            },
            "range": Array [
              43,
              57,
            ],
            "start": 43,
            "type": "AttributeValue",
            "value": "\\"dropbanner1b\\"",
          },
        },
        AttributeNode {
          "end": 119,
          "loc": Object {
            "end": Object {
              "column": 119,
              "line": 1,
            },
            "start": Object {
              "column": 58,
              "line": 1,
            },
          },
          "name": AttrNameToken {
            "end": 63,
            "loc": Object {
              "end": Object {
                "column": 63,
                "line": 1,
              },
              "start": Object {
                "column": 58,
                "line": 1,
              },
            },
            "range": Array [
              58,
              63,
            ],
            "start": 58,
            "type": "AttributeName",
            "value": "style",
          },
          "range": Array [
            58,
            119,
          ],
          "start": 58,
          "type": "Attribute",
          "value": AttrValueToken {
            "end": 119,
            "loc": Object {
              "end": Object {
                "column": 119,
                "line": 1,
              },
              "start": Object {
                "column": 64,
                "line": 1,
              },
            },
            "range": Array [
              64,
              119,
            ],
            "start": 64,
            "type": "AttributeValue",
            "value": "\\"padding-right: 20px; color: white; padding-left: 20px\\"",
          },
        },
      ],
      "children": Array [
        ElementNode {
          "attrs": Array [],
          "children": Array [
            TextNode {
              "end": 140,
              "loc": Object {
                "end": Object {
                  "column": 140,
                  "line": 1,
                },
                "start": Object {
                  "column": 126,
                  "line": 1,
                },
              },
              "range": Array [
                126,
                128,
              ],
              "start": 126,
              "type": "Text",
              "value": "웨일 사용자 목소리입니다.",
            },
          ],
          "close": CloseElementNode {
            "end": 147,
            "loc": Object {
              "end": Object {
                "column": 147,
                "line": 1,
              },
              "start": Object {
                "column": 140,
                "line": 1,
              },
            },
            "range": Array [
              140,
              147,
            ],
            "start": 140,
            "type": "CloseElement",
          },
          "end": 147,
          "loc": Object {
            "end": Object {
              "column": 147,
              "line": 1,
            },
            "start": Object {
              "column": 120,
              "line": 1,
            },
          },
          "open": OpenElementNode {
            "end": 126,
            "loc": Object {
              "end": Object {
                "column": 126,
                "line": 1,
              },
              "start": Object {
                "column": 120,
                "line": 1,
              },
            },
            "range": Array [
              120,
              126,
            ],
            "start": 120,
            "type": "OpenElement",
          },
          "range": Array [
            120,
            147,
          ],
          "selfClosing": false,
          "start": 120,
          "tagName": "span",
          "type": "Element",
        },
      ],
      "close": CloseElementNode {
        "end": 154,
        "loc": Object {
          "end": Object {
            "column": 154,
            "line": 1,
          },
          "start": Object {
            "column": 147,
            "line": 1,
          },
        },
        "range": Array [
          147,
          154,
        ],
        "start": 147,
        "type": "CloseElement",
      },
      "end": 154,
      "loc": Object {
        "end": Object {
          "column": 154,
          "line": 1,
        },
        "start": Object {
          "column": 0,
          "line": 1,
        },
      },
      "open": OpenElementNode {
        "end": 120,
        "loc": Object {
          "end": Object {
            "column": 120,
            "line": 1,
          },
          "start": Object {
            "column": 0,
            "line": 1,
          },
        },
        "range": Array [
          0,
          120,
        ],
        "start": 0,
        "type": "OpenElement",
      },
      "range": Array [
        0,
        154,
      ],
      "selfClosing": false,
      "start": 0,
      "tagName": "span",
      "type": "Element",
    },
  ],
  "comments": Array [],
  "end": 0,
  "loc": Object {
    "end": Object {
      "column": 0,
      "line": 1,
    },
    "start": Object {
      "column": 0,
      "line": 1,
    },
  },
  "range": Array [
    0,
    0,
  ],
  "start": 0,
  "tokens": Array [],
  "type": "Root",
}
`;
